plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataUpdater.java

```diff
@@ -0,0 +1,252 @@
+
+  private final Repository repository;
+
+  private final MavenFacet mavenFacet;
+
+  private final Maven2MetadataMerger metadataMerger;
+
+  private final MetadataXpp3Reader metadataReader;
+
+  private final MetadataXpp3Writer metadataWriter;
+
+  public MetadataUpdater(final boolean update, final Repository repository) {
+    this.update = update;
+    this.repository = checkNotNull(repository);
+    this.mavenFacet = repository.facet(MavenFacet.class);
+    this.metadataMerger = new Maven2MetadataMerger();
+    this.metadataReader = new MetadataXpp3Reader();
+    this.metadataWriter = new MetadataXpp3Writer();
+  }
+
+  /**
+   * Processes metadata, depending on {@link #update} value and input value of metadata parameter. If input is
+   * non-null, will update or replace depending on value of {@link #update}. If update is null, will delete if {@link
+   * #update} is {@code false}.
+   */
+  public void processMetadata(final StorageTx tx, final MavenPath metadataPath, final Maven2Metadata metadata) {
+    if (metadata != null) {
+      if (update) {
+        update(tx, metadataPath, metadata);
+      }
+      else {
+        replace(tx, metadataPath, metadata);
+      }
+    }
+    else if (!update) {
+      delete(tx, metadataPath);
+    }
+  }
+
+  /**
+   * Writes/updates metadata, merges existing one, if any.
+   */
+  public void update(final StorageTx tx, final MavenPath mavenPath, final Maven2Metadata metadata) {
+    checkNotNull(mavenPath);
+    checkNotNull(metadata);
+    try {
+      final Metadata oldMetadata = read(tx, mavenPath);
+      if (oldMetadata == null) {
+        // old does not exists, just write it
+        write(tx, mavenPath, toMetadata(metadata));
+      }
+      else {
+        // TODO: compare? unsure is it worth it, as compare would also eat CPU maybe even more that writing would
+        // update old by merging them and write out
+        final Metadata updated = metadataMerger.merge(
+            ImmutableList.of(
+                new MetadataEnvelope(repository.getName() + ":" + mavenPath.getPath(), oldMetadata),
+                new MetadataEnvelope("new:" + mavenPath.getPath(), toMetadata(metadata))
+            )
+        );
+        write(tx, mavenPath, updated);
+      }
+    }
+    catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Writes/overwrites metadata, replacing existing one, if any.
+   */
+  public void replace(final StorageTx tx, final MavenPath mavenPath, final Maven2Metadata metadata) {
+    checkNotNull(mavenPath);
+    checkNotNull(metadata);
+    try {
+      write(tx, mavenPath, toMetadata(metadata));
+    }
+    catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Deletes metadata.
+   */
+  public void delete(final StorageTx tx, final MavenPath mavenPath) {
+    checkNotNull(mavenPath);
+    try {
+      mavenFacet.delete(tx, mavenPath, mavenPath.hash(HashType.SHA1), mavenPath.hash(HashType.MD5));
```

Comment:
I suppose maven is stable enough that it's okay to hard-code the specific hashes we use here?